nengo/builder/optimizer.py

```diff
@@ -11,386 +10,707 @@
+        A_sigr[s] = Signal(data[i], name="%s[%i]" % (s.name, i), base=A)
+        assert np.all(s.initial_value == A_sigr[s].initial_value)
+        assert s.shape == A_sigr[s].shape or (
+            s.shape == () and A_sigr[s].shape == (1, 1))
+
+    reshape = operator.reshape_dot(
+        ops[0].A.initial_value, ops[0].X.initial_value,
+        ops[0].Y.initial_value, tag=ops[0].tag)
+    return (
+        operator.BsrDotInc(
+            A, X, Y, indices=indices, indptr=indptr, reshape=reshape),
+        merge_dicts(X_sigr, Y_sigr, A_sigr))
+
+
+@_OpMerger.register_op_checker(SimNeurons)
+def simneurons_checker(op1, op2):
+    return (op1.neurons == op2.neurons and
+            all(_SigMerger.check(s) for s in
+                zip(op1.all_signals, op2.all_signals)))
+
+
+def _gather(ops, key):
+    return [getattr(o, key) for o in ops]
+
+
+@_OpMerger.register_merger(SimNeurons)
+def simneurons_merger(ops):
+    J, J_sigr = _SigMerger.merge(_gather(ops, 'J'))
+    output, out_sigr = _SigMerger.merge(_gather(ops, 'output'))
+    states = []
+    states_sigr = {}
+    for signals in zip(*_gather(ops, 'states')):
+        st, st_sigr = _SigMerger.merge(signals)
+        states.append(st)
+        states_sigr.update(st_sigr)
+    return (SimNeurons(ops[0].neurons, J, output, states),
+            merge_dicts(J_sigr, out_sigr, states_sigr))
+
+
+class _OpInfo(Mapping):
+    """Analyze and store extra information about operators."""
+
+    _OpDetails = namedtuple(
+        "_OpDetails", ["first_view", "v_offset", "v_size", "v_base"])
+
+    def __init__(self):
+        self.info = {}
+
+    def __getitem__(self, op):
+        if op not in self.info:
+            try:
+                first_view = next(s for s in op.all_signals if s.is_view)
+                self.info[op] = self._OpDetails(first_view=first_view,
+                                                v_offset=first_view.offset,
+                                                v_size=first_view.nbytes,
+                                                v_base=first_view.base)
+            except StopIteration:
+                self.info[op] = self._OpDetails(
+                    first_view=None, v_offset=0, v_size=0, v_base=None)
+        return self.info[op]
+
+    def __iter__(self):
+        return iter(self.info)
+
+    def __len__(self):
+        return len(self.info)
+
+    def __repr__(self):
+        return repr(self.info)
+
+    def __str__(self):
+        return str(self.info)
+
+    def clear(self):
+        self.info.clear()
+
+
+_opinfo = _OpInfo()
+
+
+class _OpsToMerge(object):
+    """Analyze and store extra information about a list of ops to be merged."""
+
+    def __init__(self, initial_op, merged, tc):
+        self.merged = merged
+        self.tc = tc
+        self.ops = [initial_op]
+        self.optype = type(initial_op)
+
+        self.all_signals = set(initial_op.all_signals)
+        self.all_tc = set(self.tc[initial_op])
```

Comment:
Also `all_tc` makes it seem that `tc` is the transitive closure of a single operator, but it is the complete transitive closure graph (whereas `all_tc` is not a graph).